KAFKA-13888: Addition of Information in DescribeQuorumResponse about Voter Lag

niket-goel opened the pull request:
This commit adds an Admin API handler for DescribeQuorum Request and also
adds in two new fields LastFetchTimestamp and LastCaughtUpTimestamp to
the DescribeQuorumResponse as described by KIP-836.

This commit does not implement the newly added fields. Those will be
added in a subsequent commit.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

niket-goel committed:
KAFKA-13888: Addition of Information in DescribeQuorumResponse about Voter Lag

This commit adds an Admin API handler for DescribeQuorum Request and also
adds in two new fields LastFetchTimestamp and LastCaughtUpTimestamp to
the DescribeQuorumResponse as described by KIP-836.

This commit does not implement the newly added fields. Those will be
added in a subsequent commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1382,6 +1382,23 @@ default DescribeFeaturesResult describeFeatures() {
     }
 
+    /**
+     * Describe the state of the raft quorum
```
How about "metadata quorum" or "kraft quorum"? Also nit: missing period.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4181,6 +4193,61 @@ private static byte[] getSaltedPasword(ScramMechanism publicScramMechanism, byte
+                    observers.add(new ReplicaState(entry.getKey(), entry.getValue()));
+                }
+                QuorumInfo info = new QuorumInfo(topicName, leaderId, voters, observers);
+               return info;
```
nit: this looks misaligned

hachikuji commented on clients/src/main/java/org/apache/kafka/common/utils/QuorumInfo.java:
```diff
@@ -0,0 +1,90 @@
+    private final long lastFetchTimeMs;
+    private final long lastCaughtUpTimeMs;
+
+    public ReplicaState(int replicaId, long logEndOffset) {
```
Do we need to expose this?

niket-goel replied:
Sorry, I did not get what you meant. We need access to it in `KafkaAdminClient.java`. Are you suggesting it be protected? or have a builder instead of exposing the constructor? Or building it through the `QuorumInfo` itself?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/utils/QuorumInfo.java:
```diff
@@ -0,0 +1,90 @@
+      return logEndOffset;
+    }
+
+    public long lastFetchTimeMs() {
```
Can we document the result for older versions of `DescribeQuorum`? I am tempted to suggest that we change the type to `OptionalLong` to make it clear that the value may be absent for older versions.

niket-goel replied:
`OptionalLong` makes sense. Let me make that change.

hachikuji commented on clients/src/main/resources/common/message/DescribeQuorumResponse.json:
```diff
@@ -17,7 +17,7 @@
   "type": "response",
   "name": "DescribeQuorumResponse",
-  "validVersions": "0",
+  "validVersions": "0-1",
```
Usually we add some comments to describe what has changed in each version. You can see the other request specs for examples.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -318,6 +325,11 @@ public class KafkaAdminClient extends AdminClient {
+    /**
+     * The name of the internal raft metadata topic
+     */
+    private static final String METADATA_TOPIC_NAME = "__cluster_metadata";
```
Hmm, I hadn't really been thinking about the fact that we would have to expose this to the client. I guess that is the consequence of having such a general `DescribeQuorum` API. This makes me wonder if we ought to be more forward looking with the naming here. Suppose that we ultimately decide to use raft for partition replication as well. Then we might want to be able to use `DescribeQuorum` for user partitions as well, but we haven't given ourselves a lot of room for extension in the `describeQuorum` API. Would it make sense to make the new API more specific to the metadata quorum?
```java
public DescribeMetadataQuorumResult describeMetadataQuorum(DescribeMetadataQuorumOptions options)
```
It is more verbose, but it is also clearer.

We should also move this constant to `org.apache.kafka.common.internals.Topic`.

niket-goel replied:
This is a great point! I think it makes sense to give ourselves wiggle room and call this `MetadataQuorum`.
Re: moving the constant -- I was actually looking for a common place to put it so that both the clients and the server code can use the same value. Is there any class that they share? Right now the name for the topic on the server side is defined in `KafkaRaftServer`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/utils/QuorumInfo.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * This is used to describe per-partition state in the DescribeQuorumResponse.
+ */
+public class QuorumInfo {
```
Would it make sense to implement `equals` and `hashCode`? That is often useful for testing. Also it would be nice to have a good `toString` implementations.

hachikuji commented on core/src/test/scala/integration/kafka/admin/DescribeQuorumTest.scala:
```diff
@@ -0,0 +1,64 @@
+
+@Timeout(120)
+@Tag("integration")
+class DescribeQuorumTest {
```
We have another class `DescribeQuorumRequesetTest`. Would it make sense to move this test there? Perhaps we could generalize it `DescribeQuorumIntegrationTest` or something like that.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4181,6 +4193,61 @@ private static byte[] getSaltedPasword(ScramMechanism publicScramMechanism, byte
+                if (quorumResponse.data().errorCode() == Errors.NONE.code()) {
+                    future.complete(createQuorumResult(quorumResponse));
+                } else {
+                    future.completeExceptionally(Errors.forCode(quorumResponse.data().errorCode()).exception());
```
It would be helpful to have a couple unit tests in `KafkaAdminClientTest` covering failure and success cases.

hachikuji commented on raft/src/main/java/org/apache/kafka/raft/RaftConfig.java:
```diff
@@ -256,6 +256,21 @@ public static List<Node> voterConnectionsToNodes(Map<Integer, RaftConfig.Address
             .collect(Collectors.toList());
     }
 
+    public static String nodesToVoterConnections(List<Node> nodes) {
```
Seems like we're not using this anywhere. Included by mistake perhaps?

niket-goel replied:
It was included by mistake.. thanks for pointing out.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1382,6 +1382,23 @@ default DescribeFeaturesResult describeFeatures() {
+     * @param options The options to use when describing the quorum.
+     * @return The DescribeQuorumResult.
+     */
+    DescribeQuorumResult describeQuorum(DescribeQuorumOptions options);
```
By the way, usually these APIs have a default option which leaves out the `XOptions` class. For example, see `describeFeatures()` above?

Also nit: can we not split up the `describeFeatures` APIs?

hachikuji commented:
cc @dengziming for reviews

niket-goel committed:
Addressing review comments

Changes:
Added some unit tests to KafkaAdminClientTest
Merged multiple DescribeQuorumTest files
Some other refactoring

niket-goel commented:
Published another revision addressing @hachikuji's comments.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeMetadataQuorumResult.java:
```diff
@@ -0,0 +1,40 @@
+
+    private final KafkaFuture<QuorumInfo> quorumInfo;
+
+    public DescribeMetadataQuorumResult(KafkaFuture<QuorumInfo> quorumInfo) {
```
nit: Could we keep it package private?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,61 @@ void handleFailure(Throwable throwable) {
+                String topicName = response.getTopicNameByIndex(partition);
+                Integer leaderId = response.getPartitionLeaderId(topicName, partition);
+                List<ReplicaState> voters = new ArrayList<>();
+                for (Map.Entry<Integer, Long> entry: response.getVoterOffsets(topicName, partition).entrySet()) {
```
nit: We could use `response.getVoterOffsets(topicName, partition).forEach((replicaId, logEndOffset) ->`. That makes the code a bit more readable in my opinion.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,61 @@ void handleFailure(Throwable throwable) {
+                Integer leaderId = response.getPartitionLeaderId(topicName, partition);
+                List<ReplicaState> voters = new ArrayList<>();
+                for (Map.Entry<Integer, Long> entry: response.getVoterOffsets(topicName, partition).entrySet()) {
+                    voters.add(new ReplicaState(entry.getKey(), entry.getValue(), OptionalLong.empty(), OptionalLong.empty()));
```
hm.. `OptionalLong.empty()` seems wrong here. Don't you need to put the last fetch time and the last caughtup time?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,61 @@ void handleFailure(Throwable throwable) {
+                }
+                List<ReplicaState> observers = new ArrayList<>();
+                for (Map.Entry<Integer, Long> entry: response.getObserverOffsets(topicName, partition).entrySet()) {
+                    observers.add(new ReplicaState(entry.getKey(), entry.getValue(), OptionalLong.empty(), OptionalLong.empty()));
```
Same question here.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,61 @@ void handleFailure(Throwable throwable) {
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                completeAllExceptionally(Collections.singletonList(future), throwable);
```
nit: Why not directly calling `future.completeExceptionally`?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,61 @@ void handleFailure(Throwable throwable) {
+            private QuorumInfo createQuorumResult(final DescribeQuorumResponse response) {
+                Integer partition = 0;
+                String topicName = response.getTopicNameByIndex(partition);
+                Integer leaderId = response.getPartitionLeaderId(topicName, partition);
```
Implementation of `getPartitionLeaderId`, `getVoterOffsets` and `getObserverOffsets` always start by looking up the relevant topic/partition in the response. Did you consider doing this lookup once and use the result to construct the `QuorumInfo`?

hachikuji replied:
I agree it is a little awkward. A common way of structuring responses which include topic partitions is to first build something like `Map<TopicPartition, QuorumInfo>`. Then we can pluck out the partitions we need. For example, see `logDirDescriptions` in this class.

hachikuji replied:
@niket-goel any comment here? I think this part still reads awkward. Converting to an intermediate map is conventional. An alternative would be to do a quick validation of the response. We can structure the checks like this:

1. Check top-level error code
2. Verify only one topic in the response which matches metadata topic
3. Verify only one partition in the response with id 0
4. Check partition-level error code.

This is similar to how we handle the request in `KafkaRaftClient.handleDescribeQuorumRequest`.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,61 @@ void handleFailure(Throwable throwable) {
+        final KafkaFutureImpl<QuorumInfo> future = new KafkaFutureImpl<>();
+        final long now = time.milliseconds();
+        final Call call = new Call(
+                "describeQuorum", calcDeadlineMs(now, options.timeoutMs()), provider) {
```
nit: Should we use `describeMetadataQuorum`?

dajac commented on clients/src/main/java/org/apache/kafka/common/utils/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common;
```
Is `org.apache.kafka.common` the right package here? It seems preferable to put them in `org.apache.kafka.clients.admin` package.

dajac commented on clients/src/main/resources/common/message/DescribeQuorumResponse.json:
```diff
@@ -17,9 +17,10 @@
+  "validVersions": "0-1",
   "flexibleVersions": "0+",
   "fields": [
+    // Version 1 adds LastFetchTimeStamp and LastCaughtUpTimestamp in ReplicaState
```
nit: We usually put this before `validVersions`.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -592,6 +597,26 @@ private static ApiVersionsResponse prepareApiVersionsResponseForDescribeFeatures
+        }
+        return new DescribeQuorumResponse(
+                new DescribeQuorumResponseData()
+                .setErrorCode(error.code()));
```
nit: Indentation seems off here.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -592,6 +597,26 @@ private static ApiVersionsResponse prepareApiVersionsResponseForDescribeFeatures
+        if (error == Errors.NONE) {
+            return new DescribeQuorumResponse(DescribeQuorumResponse.singletonResponse(
+                        new TopicPartition(Topic.METADATA_TOPIC_NAME, 0),
+                        0, 0, 0,
+                        singletonList(new DescribeQuorumResponseData.ReplicaState()),
+                        singletonList(new DescribeQuorumResponseData.ReplicaState()))
```
It would be better to fully populate the response here. That would have caught the issue that I mentioned earlier.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4846,6 +4871,32 @@ public void testDescribeFeaturesFailure() {
+                    body -> body instanceof DescribeQuorumRequest,
+                    prepareDescribeQuorumResponse(Errors.INVALID_REQUEST));
+            final KafkaFuture<QuorumInfo> future = env.adminClient().describeMetadataQuorum().quorumInfo();
+            final ExecutionException e = assertThrows(ExecutionException.class, future::get);
+            assertEquals(e.getCause().getClass(), Errors.INVALID_REQUEST.exception().getClass());
```
nit: You could use `TestUtils.assertFutureThrows`.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumIntegrationTest.scala:
```diff
@@ -80,6 +85,36 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
   }
 
+  @ClusterTest
+  def testDescribeQuorumRequestToBrokers() = {
```
Looking at the other tests, this test suite does not seem to be the right place for this test. I wonder if `KRaftClusterTest` would be more appropriate for instance but I don't feel strong about this.

niket-goel replied:
I actually had this test in a separate suite akin to `KRaftClusterTest` before, but moved it here because Jason pointed out that a test file for this API already existed. I think I am going to move the test to where you suggested. It does make more sense to me there.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumIntegrationTest.scala:
```diff
@@ -17,25 +17,30 @@
+  val log = LoggerFactory.getLogger(classOf[DescribeQuorumIntegrationTest])
 
   @ClusterTest(clusterType = Type.ZK)
   def testDescribeQuorumNotSupportedByZkBrokers(): Unit = {
```
Should we extend `testDescribeQuorum` to verify that new fields are set when v1 is used and not set when v0 is used?

niket-goel replied:
Thanks for the detailed review @dajac. Some of the misses were me working with multiple versions of the code. Appreciate you catching those. I will udpate the next version with all of your comments addressed. One quick question about this comment though:
Is there an example of this somewhere that I can look at? Not sure how to change and verify the version of the request.

niket-goel replied:
Another issue is that the new fields are defaulted to `-1` when not set, but also `-1` when the value for a particualr `voter` or `observer` is unknown. Thinking if there is a way test this reliably.

dajac reviewed (commented):
@niket-goel Thanks for the PR. I left a few comments.

dengziming reviewed (commented):
Thanks for this work @niket-goel ,  I have 2 suggestions before going into details. Firstly, we can add a test in PlaintextAdminIntegrationTest for this change, the second one is less important, it's recommended to use `AdminApiHandler` instead of using `runnable.call` directly, you can refer to FenceProducersHandler.java as a simple example.

niket-goel committed:
Addressing review comments

Changes include:
* Fixing fields not being correctly set when processing API response in
  Admin Client
* Modified test to cover the above case
* Test refactoring
* Minor fixes in naming and indentation

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1446,6 +1446,35 @@ default DescribeFeaturesResult describeFeatures() {
+     *
+     * @return the {@link DescribeMetadataQuorumResult} containing the result
+     */
+    default DescribeMetadataQuorumResult describeMetadataQuorum() {
```
It is very common for API details to change during implementation. Once we're satisfied and ready to merge, we should update the KIP. Typically we would also send a message to the vote thread in case there are any concerns.

niket-goel replied:
I will update the KIP and the thread once this PR is approved to reflect the changes here.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+
+    @Override
+    public String toString() {
+        return "QuorumInfo{" +
```
nit: conventionally (if perhaps not always consistently), we prefer parenthesis for `toString` implementations

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+        private final OptionalLong lastFetchTimeMs;
+        private final OptionalLong lastCaughtUpTimeMs;
+
+        public ReplicaState() {
```
nit: do we need to expose this? Usually we only expose the minimum necessary constructors since we're often stuck with what we expose for a long time.

niket-goel replied:
~~I need the constructor access in `DescribeQuorumResponse.java`. I can sort of work around that by returning the fields raw or encapsulated in a different object, but i thought this was simpler. If we want to prioritize hiding the constructor more, I can change it so that the `getVoterInfo` and `getObserverInfo` return raw fields. Do you have any thoughts on this?~~

niket-goel replied:
Never mind. I think I was just being a little stupid :)

hachikuji commented on clients/src/main/resources/common/message/DescribeQuorumRequest.json:
```diff
@@ -18,7 +18,7 @@
   "listeners": ["broker", "controller"],
   "name": "DescribeQuorumRequest",
-  "validVersions": "0",
+  "validVersions": "0-1",
```
nit: maybe we can something like this:
```
  // Version 1 adds additional fields in the response. The request is unchanged.
```

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeQuorumResponse.java:
```diff
@@ -93,4 +97,57 @@ public static DescribeQuorumResponseData singletonResponse(TopicPartition topicP
+        return voterInfo;
+    }
+
+    public List<QuorumInfo.ReplicaState> getObserverInfo(String topicName, Integer partition) {
```
nit: `getVoterInfo` and `getObserverInfo` seem basically the same. Feels like we can factor out the common logic.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4327,58 @@ void handleFailure(Throwable throwable) {
+            @Override
+            void handleResponse(AbstractResponse response) {
+                final DescribeQuorumResponse quorumResponse = (DescribeQuorumResponse) response;
+                if (quorumResponse.data().errorCode() == Errors.NONE.code()) {
```
Hmm.. We're checking the top-level error code, but the response also has a partition-level error to check.

hachikuji commented on core/src/test/scala/unit/kafka/server/DescribeQuorumIntegrationTest.scala:
```diff
@@ -26,16 +25,17 @@ import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 @ClusterTestDefaults(clusterType = Type.KRAFT)
 @Tag("integration")
-class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+class DescribeQuorumTest(cluster: ClusterInstance) {
```
I'm a little confused what happened here, but it looks like we should get rid of this class and rename `DescribeQuorumTest` to `DescribeQuorumIntegrationTest`. Also, did we lose the admin integration test you added?

niket-goel replied:
I changed it so that the new test resides in `KRaftClusterTest`. This was Daivd's suggestion and I felt it made more sense there, especially because that new test used a different test framework than these ones.
The change in the name got included by mistake and I will revert it what it was.

dengziming commented on core/src/test/scala/unit/kafka/server/DescribeQuorumIntegrationTest.scala:
```diff
@@ -26,16 +25,17 @@ import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 @ClusterTestDefaults(clusterType = Type.KRAFT)
 @Tag("integration")
-class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+class DescribeQuorumTest(cluster: ClusterInstance) {
```
The class name and file name are inconsistent.

dengziming commented on core/src/test/scala/unit/kafka/server/DescribeQuorumTest.scala:
```diff
@@ -0,0 +1,106 @@
+@ExtendWith(value = Array(classOf[ClusterTestExtensions]))
+@ClusterTestDefaults(clusterType = Type.KRAFT)
+@Tag("integration")
+class DescribeQuorumIntegrationTest(cluster: ClusterInstance) {
```
Ditto, The class name and file name are inconsistent.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4327,58 @@ void handleFailure(Throwable throwable) {
+
+            private QuorumInfo createQuorumResult(final DescribeQuorumResponse response) {
+                Integer partition = 0;
+                String topicName = response.getTopicNameByIndex(partition);
```
`response.getTopicNameByIndex(partition)` is a little confusing here, how can we get a topic by partition, I think we should rename partition to topicIndex or use `response.getTopicNameByIndex(0)` directly.

dengziming commented on core/src/test/scala/unit/kafka/server/DescribeQuorumTest.scala:
```diff
@@ -0,0 +1,106 @@
+  ): T = {
+    IntegrationTestUtils.connectAndReceive(
+      request,
+      cluster.brokerSocketServers().asScala.head,
```
I wonder if we can send it to controllerSockerServer directly? for example, we set bootstrap.server=localhost:9093, we should prevent a client from doing this.

dengziming commented on core/src/test/scala/unit/kafka/server/DescribeQuorumTest.scala:
```diff
@@ -0,0 +1,106 @@
+        .getOrElse(throw new AssertionError("Failed to find leader among current voter states"))
+        assertTrue(leaderState.logEndOffset > 0)
+
+        val voterData = partitionData.currentVoters().asScala
```
It's worth checking `partitionData.observers()` here too.

dengziming commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4846,6 +4878,31 @@ public void testDescribeFeaturesFailure() {
+    @Test
+    public void testDescribeMetadataQuorumSuccess() throws Exception {
+        try (final AdminClientUnitTestEnv env = mockClientEnv()) {
+            env.kafkaClient().setNodeApiVersions(NodeApiVersions.create((short) 55, (short) 0, (short) 1));
```
ApiKeys.DESCRIBE_QUORUM.id,  DESCRIBE_QUORUM.highestSupportVersion， DESCRIBE_QUORUM.lowestSupportVersion may be better here.

dengziming reviewed (commented):
Thanks for the update @niket-goel , left some comments.

niket-goel committed:
Addressing more feedback from the review

niket-goel commented:
Thanks for review @dengziming .

A few comments:
> Firstly, we can add a test in PlaintextAdminIntegrationTest for this change

I looked at the file and the tests there look similar to the what was added to `KraftClusterTest`. Could you check if the intent was to run an integration test like that? We can always increase the checks that test does.

> the second one is less important, it's recommended to use AdminApiHandler instead of using runnable.call directly, you can refer to FenceProducersHandler.java as a simple example.

I looked at the class and it looks like a wrapper for the runnable call. Do you mind if we make that improvement in a subsequent change (just trying to get this PR under control :) )

> I wonder if we can send it to controllerSockerServer directly? for example, we set bootstrap.server=localhost:9093, we should prevent a client from doing this.

A couple of things here -- This code existed before this change (I just moved it). I would love to improve it though. I just did not understand what exactly your comment meant. Could you please elaborate?

Apart from the above questions I think I have addressed all the other concerns everyone raised in the latest version.

dengziming commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -725,4 +725,34 @@ class KRaftClusterTest {
+        val quorumInfo = quorumState.quorumInfo().get()
+
+        assertEquals(KafkaRaftServer.MetadataTopic, quorumInfo.topic())
+        assertEquals(3, quorumInfo.voters.size())
```
we can also check observers here.

dengziming commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -725,4 +725,34 @@ class KRaftClusterTest {
   }
+
+  @Test
+  def testDescribeQuorumRequestToBrokers() = {
```
nit: add : Unit after method

dengziming commented:
@niket-goel Thanks for the reply, I left 2 minor comments. I still find the class name is `DescribeQuorumIntegrationTest` but the file name is `DescribeQuorumTest.scala`, you should make them the same, I run your newly added tests locally but some of them failed, maybe there are still some bugs to be fixed.

For the last problem,  I mean we can also get quorum directly from the controller, I can give you an example:

For a single node KRaft cluster, we can get quorum info from BrokerServer:
```
properties.put("bootstrap.servers",  "localhost:9092"); // connect to BrokerServer 9092
AdminClient client = AdminClient.create(Properties)
client.describeMetadataQuorum()
```

We can also connect to ControllerServer:
```
properties.put("bootstrap.servers",  "localhost:9093"); // connect to ControllerServer 9093
AdminClient client = AdminClient.create(Properties)
client.describeMetadataQuorum()
```

We can leave this problem to a separate PR since it's unrelated here.

niket-goel committed:
Fixing some minor comments

niket-goel commented:
Thanks @dengziming . I have pushed another version where all tests are passing locally. Appreciate your help with the review.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeQuorumResponse.java:
```diff
@@ -93,4 +94,57 @@ public static DescribeQuorumResponseData singletonResponse(TopicPartition topicP
+        if (topic != null) {
+            List<ReplicaState> replicaStates = getVoterInfo ? topic.partitions().get(partition).currentVoters()
+                    : topic.partitions().get(partition).observers();
+            //replicaStates.forEach(
```
Left in by mistake?

niket-goel committed:
minor: removing some unused code

hachikuji commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +56,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+
+      val leaderState = partitionData.currentVoters.asScala.find(_.replicaId == leaderId)
+        .getOrElse(throw new AssertionError("Failed to find leader among current voter states"))
+        assertTrue(leaderState.logEndOffset > 0)
```
nit: these are all misaligned

hachikuji commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +56,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+        val voterData = partitionData.currentVoters().asScala
+        val observerData = partitionData.observers().asScala
+        if (version == 0) {
+          voterData.foreach( state => {
```
nit: the idiomatic way to write this is 
```scala
voterData.foreach { state =>
  ...
}
```

hachikuji commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +56,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+        val observerData = partitionData.observers().asScala
+        if (version == 0) {
+          voterData.foreach( state => {
+            assertTrue(state.lastFetchTimestamp() == -1)
```
nit: we should use `assertEquals`. The advantage is that we can see what the actual value was in the failure message, which is sometimes useful to understand the failure.

hachikuji commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +56,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
-    val leaderState = partitionData.currentVoters.asScala.find(_.replicaId == leaderId)
-      .getOrElse(throw new AssertionError("Failed to find leader among current voter states"))
-    assertTrue(leaderState.logEndOffset > 0)
+    for (version <- ApiKeys.DESCRIBE_QUORUM.oldestVersion to ApiKeys.DESCRIBE_QUORUM.latestVersion) {
```
How about this?
```scala
for (version <- ApiKeys.DESCRIBE_QUORUM.allVersions.asScala) {
```

hachikuji commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -26,16 +25,19 @@ import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 @ClusterTestDefaults(clusterType = Type.KRAFT)
 @Tag("integration")
 class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+  val log = LoggerFactory.getLogger(classOf[DescribeQuorumRequestTest])
```
Seems unused?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4330,73 @@ void handleFailure(Throwable throwable) {
+            }
+
+            @Override
+            DescribeQuorumRequest.Builder createRequest(int timeoutMs) {
```
nit: it doesn't look like `timeoutMs` is used? Also, could we use `DescribeQuorumRequest.singletonRequest`?

hachikuji replied:
You may have missed the comment here.

hachikuji replied:
Nevermind, this is an override.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4846,6 +4878,35 @@ public void testDescribeFeaturesFailure() {
+    }
+
+    @Test
+    public void testDescribeMetadataQuorumFailure() {
```
Could we have a test case with a partition-level error?

hachikuji reviewed (commented):
Thanks for the updates. Left a few more comments.

dengziming commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -725,4 +725,35 @@ class KRaftClusterTest {
+
+        assertEquals(KafkaRaftServer.MetadataTopic, quorumInfo.topic())
+        assertEquals(3, quorumInfo.voters.size())
+        assertEquals(0, quorumInfo.observers.size())
```
I wonder why there isn't an observer, we should have 4 observers since we have 4 brokers, this may also not be related to this PR, I will spend some time investigating.

dengziming replied:
I find the cause here, We set nodeId=-1 if it's a broker so observers.size==0
https://github.com/apache/kafka/blob/4c9eeef5b2dff9a4f0977fbc5ac7eaaf930d0d0e/core/src/main/scala/kafka/raft/RaftManager.scala#L185-L189

I changed it to `val nodeId = OptionalInt.of(config.nodeId)`, then observers.size==4

dajac replied:
Do we need to do something about this in this PR?

dajac replied:
This comment has not been addressed.

dajac replied:
@hachikuji Is it the expected behavior to have zero observers?

niket-goel committed:
Addressing comments and improving DescribeMetadatga Quorum API handling

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/Admin.java:
```diff
@@ -1446,6 +1446,35 @@ default DescribeFeaturesResult describeFeatures() {
+    /**
+     * Describes the state of the metadata quorum.
+     * <p>
+     * This is a convenience method for {@link #describeMetadataQuorum(DescribeMetadataQuorumOptions)}  with default options.
```
nit: There is an extra space before `with`.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeMetadataQuorumOptions.java:
```diff
@@ -0,0 +1,25 @@
+
+/**
+ * Options for {@link ConfluentAdmin#describeQuorum(DescribeMetadataQuorumOptions)}.
+ *
```
nit: We could remove this empty line.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeMetadataQuorumResult.java:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * The result of {@link Admin#describeMetadataQuorum(DescribeMetadataQuorumOptions)}
+ *
```
nit: We could remove this empty line.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/DescribeMetadataQuorumResult.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     * Returns a future QuorumInfo
```
nit: Should we say `Returns a future containing the quorum info.`?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,91 @@ void handleFailure(Throwable throwable) {
+                    if (quorumResponse.data().topics().size() > 1) {
+                        log.error("DescribeMetadataQuorum received {} topics when 1 was expected",
+                                quorumResponse.data().topics().size());
+                        throw new UnknownServerException();
```
* We should add a message in every exception provided back to the user. If we do this, we can log as debug instead of error.
* Instead of throwing the exception, we usually complete the future and return.

niket-goel replied:
Good point about adding the message to the exception being returned.

The behavior of the code block is equivalent to completing the future and returning. I just didn't want to repeat the future completion code. Is there a well known convention to handle code blocks like this in JAVA?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,91 @@ void handleFailure(Throwable throwable) {
+                    if (partition.errorCode() != Errors.NONE.code()) {
+                        throw Errors.forCode(partition.errorCode()).exception();
+                    }
+                    future.complete(createQuorumResult(quorumResponse));
```
At this point, we have the `partition` that we are interested in. Couldn't we directly use it to populate `QuorumInfo` instead of looking it up again in `createQuorumResult`?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+    private final List<ReplicaState> voters;
+    private final List<ReplicaState> observers;
+
+    public QuorumInfo(String topic, Integer leaderId, List<ReplicaState> voters, List<ReplicaState> observers) {
```
nit: Could we keep it package private?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+import java.util.OptionalLong;
+
+/**
+ * This is used to describe per-partition state in the DescribeQuorumResponse.
```
The reference to DescribeQuorumResponse is a bit weird here for our users. Could we to better describe what QuorumInfo represents?

hachikuji replied:
Yeah, I agree with this. Maybe we can just say that this class contains useful debugging state for KRaft replication.

dajac replied:
This comment may have been missed.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+            this.lastCaughtUpTimeMs = lastCaughtUpTimeMs;
+        }
+
+        public int replicaId() {
```
nit: Should we add javadoc to all the getters to be consistent? It is a bit weird to have for some but not for all.

dajac commented on clients/src/main/resources/common/message/DescribeQuorumRequest.json:
```diff
@@ -18,7 +18,8 @@
   "listeners": ["broker", "controller"],
   "name": "DescribeQuorumRequest",
-  "validVersions": "0",
+  // Version 1 adds additional fields in the response. The request is unchanged.
```
nit: Could we add the KIP reference?

dajac commented on clients/src/main/resources/common/message/DescribeQuorumResponse.json:
```diff
@@ -17,7 +17,8 @@
   "type": "response",
   "name": "DescribeQuorumResponse",
-  "validVersions": "0",
+  // Version 1 adds LastFetchTimeStamp and LastCaughtUpTimestamp in ReplicaState
```
nit: Could we add the KIP reference?

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4833,6 +4879,92 @@ public void testDescribeFeaturesFailure() {
+    @Test
+    public void testDescribeMetadataQuorumSuccess() throws Exception {
+        try (final AdminClientUnitTestEnv env = mockClientEnv()) {
+            env.kafkaClient().setNodeApiVersions(NodeApiVersions.create(ApiKeys.DESCRIBE_QUORUM.id,
```
nit: You could use `NodeApiVersions.create()`.

niket-goel replied:
Not sure about this comment. I think I would need to add the `DESCRIBE_QUORUM` API to the list of known APIs even if we used the default constructor, right? If I do not do that, the test encounters an error where it claims to not know about the `DESCRIBE_QUORUM` API.

dengziming replied:
We can't use `NodeApiVersions.create()` here due to a bug here, which I'm trying to fix here: https://github.com/apache/kafka/pull/11784

dajac replied:
Ack. I was not aware of this bug.

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4833,6 +4879,92 @@ public void testDescribeFeaturesFailure() {
+                    body -> body instanceof DescribeQuorumRequest,
+                    prepareDescribeQuorumResponse(Errors.INVALID_REQUEST, Errors.NONE, false, false, false, false));
+            KafkaFuture<QuorumInfo> future = env.adminClient().describeMetadataQuorum().quorumInfo();
+            TestUtils.assertFutureThrows(future, Errors.INVALID_REQUEST.exception().getClass());
```
nit: `InvalidRequestException.class`?

dajac commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4833,6 +4879,92 @@ public void testDescribeFeaturesFailure() {
+            TestUtils.assertFutureThrows(future, Errors.INVALID_REQUEST.exception().getClass());
+        }
+    }
+
```
Could we add a test case where `lastFetchTimestamp` and `lastCaughtUpTimestamp` are not provided in the response? I suppose that should result in empty options, right?

dajac commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -778,4 +778,35 @@ class KRaftClusterTest {
+        val quorumState = admin.describeMetadataQuorum(new DescribeMetadataQuorumOptions)
+        val quorumInfo = quorumState.quorumInfo().get()
+
+        assertEquals(KafkaRaftServer.MetadataTopic, quorumInfo.topic())
```
Should we verify all the fields to ensure that the server give us the info that we expected?

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
-    val leaderState = partitionData.currentVoters.asScala.find(_.replicaId == leaderId)
-      .getOrElse(throw new AssertionError("Failed to find leader among current voter states"))
-    assertTrue(leaderState.logEndOffset > 0)
+    for (version <- ApiKeys.DESCRIBE_QUORUM.allVersions.asScala) {
```
I wonder if we could use `@ApiKeyVersionsSource(apiKey = ApiKeys.DESCRIBE_QUORUM)`. I don't know if that works with `@ClusterTest` though.

hachikuji replied:
I had considered giving the same comment, but I thought it might be better--or at least faster--to use the same integration test setup.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+    for (version <- ApiKeys.DESCRIBE_QUORUM.allVersions.asScala) {
+      val request = new DescribeQuorumRequest.Builder(
+        singletonRequest(KafkaRaftServer.MetadataPartition)
+      ).build(version.asInstanceOf[Short])
```
nit: We could use `.toShort`.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+      val leaderId = partitionData.leaderId
+      assertTrue(leaderId > 0)
+
+      val leaderState = partitionData.currentVoters.asScala.find(_.replicaId == leaderId)
```
Should we assert all the fields of all the replicas? I would also verify that we have the correct voters and observers, etc.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+        .getOrElse(throw new AssertionError("Failed to find leader among current voter states"))
+      assertTrue(leaderState.logEndOffset > 0)
+
+        val voterData = partitionData.currentVoters().asScala
```
nit: Indentation is off here.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,44 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+
+        val voterData = partitionData.currentVoters().asScala
+        val observerData = partitionData.observers().asScala
+        if (version == 0) {
```
Could we verify anything if version is > 0?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+ * This is used to describe per-partition state in the DescribeQuorumResponse.
+ */
+public class QuorumInfo {
+    private final String topic;
```
It is curious that we have the topic but not the partition. Is it on purpose, perhaps because we assume only one partition anyway?

dengziming replied:
I think It's better to add partition here because we are making way for multi-raft.

niket-goel replied:
The intention was to keep it simple for now (given we are not using the system as multi-raft yet). We could add in the partition information here as well, but then how far do we want to go. Do we want this structure to mirror the `TopicData` and `PartitionData` types in the `DescribeQuorumResponseData`, or do we just want to do something simple (like adding in a partition ID) for now?

hachikuji replied:
@niket-goel Maybe we had better leave topic/partition out of this object? I don't think there's any reason to expose `__cluster_metadata` to users of the admin client.

dajac reviewed (commented):
@niket-goel Thanks for the update. I made another pass on it and left some comments.

niket-goel committed:
Addressing more feedback. Adding some missing test scenarios and fixing minor typos and missing comments

niket-goel committed:
Minor: removing a refactored field from a test

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,94 @@ void handleFailure(Throwable throwable) {
+                    }
+                    future.complete(createQuorumResult(partition));
+                } catch (RuntimeException e) {
+                    throw e;
```
Is the intent to retry here? For some errors, such as auth failures, we probably would rather fail fast.

niket-goel replied:
The reason I added this block is because I noticed a gradle warning which suggested that (with the addition of the general Exception catch block), some runtime exceptions might get hidden. A little reading on this suggested that a best practice is to catch and re-throw runtime exceptions. I guess your comment here is that it is best to just do the same thing with runtime exceptions as we are doing with other exceptions and complete the future with an error. Am I understanding that correctly?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,94 @@ void handleFailure(Throwable throwable) {
+                    future.complete(createQuorumResult(partition));
+                } catch (RuntimeException e) {
+                    throw e;
+                } catch (Exception e) {
```
Which exceptions are we catching here?

niket-goel replied:
So this block is me trying to have a single `future.completeExceptionally()` call in this code block. We are catching `UnknownServerException` and any exception that might be returned by the server in its response. The generic catch handler can be avoided by not throwing the exceptions returned by the server, and instead just completing exceptionally within the block above. I just find it more maintainable to have single calls to completions etc. If this is creating some other issues in the code, i can change this.

hachikuji replied:
`UnknownServerException` extends `KafkaException`, which extends `RuntimeException`. So I think all the errors that we are raising above get re-thrown in the previous `catch`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeQuorumResponse.java:
```diff
@@ -93,4 +94,25 @@ public static DescribeQuorumResponseData singletonResponse(TopicPartition topicP
+     * @param getVoterInfo Return the voter information if true, return observers otherwise
+     * @return List of {@link ReplicaState}
+     */
+    private List<ReplicaState> getReplicaInfo(String topicName, Integer partition, boolean getVoterInfo) {
```
nit: are we using this anymore?

niket-goel replied:
no we are not. Thanks for the catch!

hachikuji commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -778,4 +778,35 @@ class KRaftClusterTest {
+        admin.close()
+      }
+    } finally {
+      cluster.close
```
nit: we usually add parenthesis for mutators

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4337,25 +4337,22 @@ public DescribeMetadataQuorumResult describeMetadataQuorum(DescribeMetadataQuoru
                     voters.add(new QuorumInfo.ReplicaState(v.replicaId(),
                                 v.logEndOffset(),
                                 OptionalLong.of(v.lastFetchTimestamp()),
                                 OptionalLong.of(v.lastCaughtUpTimestamp())));
```
When `lastFetchTimestamp` or `lastCaughtUpTimestamp` are not provided (equals to -1), don't we want to return an empty option instead of returning an option containing -1?

niket-goel replied:
I went back and forth between that and ended up returning a -1 optional here. I now remember that the original intention was to have an empty optional. Will address this.

niket-goel committed:
More review feedback

* Removed explicit exception handling from DescribeMetadataQuorum::handleResponse.
* Removed an unused method
* Fixed Api handler to return OptionalEmpty for unknown fields.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,88 @@ void handleFailure(Throwable throwable) {
+                if (quorumResponse.data().errorCode() != Errors.NONE.code()) {
+                    throw Errors.forCode(quorumResponse.data().errorCode()).exception();
+                }
+                if (quorumResponse.data().topics().size() > 1) {
```
Maybe we should check if size is not equal to 1 here and below. I guess an empty list is also possible.

niket-goel committed:
strengthening a check

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,88 @@ void handleFailure(Throwable throwable) {
+                            v.lastCaughtUpTimestamp() == -1 ? OptionalLong.empty() : OptionalLong.of(v.lastCaughtUpTimestamp())));
+                });
+                partition.observers().forEach(o -> {
+                    observers.add(new QuorumInfo.ReplicaState(o.replicaId(),
+                            o.logEndOffset(),
+                            o.lastFetchTimestamp() == -1 ? OptionalLong.empty() : OptionalLong.of(o.lastFetchTimestamp()),
+                            o.lastCaughtUpTimestamp() == -1 ? OptionalLong.empty() : OptionalLong.of(o.lastCaughtUpTimestamp())));
```
Could we extract this bloc into an helper method and share it with the voters part?

niket-goel replied:
Do we really think this is worth another method.?

dajac replied:
Yes. We should avoid code duplication. Btw, you could also use the stream api here.: `partition.observers().stream.map(function).collect(Collectors.toList())`. I leave this one up to you.

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4321,6 +4328,88 @@ void handleFailure(Throwable throwable) {
+            void handleResponse(AbstractResponse response) {
+                final DescribeQuorumResponse quorumResponse = (DescribeQuorumResponse) response;
+                if (quorumResponse.data().errorCode() != Errors.NONE.code()) {
+                    throw Errors.forCode(quorumResponse.data().errorCode()).exception();
```
For my understanding, is the idea to retry the `Retriable` ones here? I also suppose that the `UnknownServerException` thrown will be caught and result in calling `handleFailure`. Am I right?

niket-goel replied:
That is correct understanding. The intent (based on some discussion with Jason) is to allow the default handler to manage exceptions returned by the server. The `UnknownServerException` that we raise here will end up calling `handleFailure` and complete exceptionally.

dajac replied:
Great, thanks.

dajac commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -778,4 +778,35 @@ class KRaftClusterTest {
+        val quorumState = admin.describeMetadataQuorum(new DescribeMetadataQuorumOptions)
+        val quorumInfo = quorumState.quorumInfo().get()
+
+        assertEquals(0, quorumInfo.leaderId())
```
Should we check all the data in quorum info in this integration test? Just checking the leader and the number of voters/observers seems weak to me.

niket-goel replied:
Done!

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,46 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+
+      val voterData = partitionData.currentVoters().asScala
+      val observerData = partitionData.observers().asScala
+      if (version == 0) {
```
What about version > 0?

niket-goel replied:
I think I missed replying to this comment earlier. Apologies about that. I am not sure if we can check anything for versions greater than zero at the moment. As of this PR we are not setting the fields in the response and so we cannot verify that. Was there something you had in mind that we should verify for versions > 0?

dajac replied:
In this case, should we just remove that version check and verify all versions? My understanding is that they should all be the same at the moment. We can update the test when we implement the server side. What do you think?

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,46 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+      val observerData = partitionData.observers().asScala
+      if (version == 0) {
+        voterData.foreach { state =>
+          assertTrue(0 < state.replicaId)
```
Should we also assert `logEndOffset`?

niket-goel replied:
yes.

niket-goel committed:
Fxing a typo in the comments

niket-goel committed:
minor: additional test checks

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,149 @@
+            this(0, 0, OptionalLong.empty(), OptionalLong.empty());
+        }
+
+        ReplicaState(int replicaId, long logEndOffset,
+                OptionalLong lastFetchTimeMs, OptionalLong lastCaughtUpTimeMs) {
```
nit: The code format seems a bit off here. I think that we would format like this:

```
ReplicaState(
    int replicaId,
    long logEndOffset,
    OptionalLong lastFetchTimeMs,
    OptionalLong lastCaughtUpTimeMs
) {
```

dajac commented on clients/src/main/java/org/apache/kafka/common/requests/DescribeQuorumResponse.java:
```diff
@@ -93,4 +93,5 @@ public static DescribeQuorumResponseData singletonResponse(TopicPartition topicP
     public static DescribeQuorumResponse parse(ByteBuffer buffer, short version) {
         return new DescribeQuorumResponse(new DescribeQuorumResponseData(new ByteBufferAccessor(buffer), version));
     }
+
```
nit: This empty line could be removed.

dajac commented on clients/src/main/resources/common/message/DescribeQuorumResponse.json:
```diff
@@ -17,7 +17,8 @@
   "type": "response",
   "name": "DescribeQuorumResponse",
-  "validVersions": "0",
+  // Version 1 adds LastFetchTimeStamp and LastCaughtUpTimestamp in ReplicaState (KIP-836)
```
nit: Can we add a dot at the end of the sentence?

dajac commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -778,4 +778,43 @@ class KRaftClusterTest {
+        val quorumState = admin.describeMetadataQuorum(new DescribeMetadataQuorumOptions)
+        val quorumInfo = quorumState.quorumInfo().get()
+
+        assertEquals(3000, quorumInfo.leaderId())
```
nit: We usually omit parenthesis for getters. There are a few other cases in this file.

dajac commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -778,4 +778,43 @@ class KRaftClusterTest {
+        assertEquals(3000, quorumInfo.leaderId())
+        assertEquals(0, quorumInfo.observers.size())
+        assertEquals(3, quorumInfo.voters.size())
+        quorumInfo.voters().forEach( voter => {
```
nit: We usually format block as follow: `quorumInfo.voters.forEach { voter =>`.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,48 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+      assertTrue(leaderState.logEndOffset > 0)
+
+      val voterData = partitionData.currentVoters().asScala
+      val observerData = partitionData.observers().asScala
```
nit: Should we assert the number of voters/observers as well?

dajac reviewed (commented):
@niket-goel Thanks for the update. I left a few more minor comments. There are also a few previous comments which have not been addressed.

niket-goel committed:
Tightening up the code a bit based on some more feedback

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,153 @@
+        }
+
+        ReplicaState(
+                int replicaId,
+                long logEndOffset,
+                OptionalLong lastFetchTimeMs,
+                OptionalLong lastCaughtUpTimeMs
```
nit: We usually use 4 spaces indentation here.

dajac commented on core/src/test/scala/integration/kafka/server/KRaftClusterTest.scala:
```diff
@@ -778,4 +778,44 @@ class KRaftClusterTest {
+        val quorumInfo = quorumState.quorumInfo.get()
+
+        assertEquals(0, quorumInfo.observers.size())
+        assertEquals(3, quorumInfo.voters.size())
```
nit: We could omit parenthesis after `size`. The same for the previous line.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,48 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+        .getOrElse(throw new AssertionError("Failed to find leader among current voter states"))
+      assertTrue(leaderState.logEndOffset > 0)
+
+      val voterData = partitionData.currentVoters().asScala
```
nit: We could omit parenthesis after `currentVoters`. The same for the next line.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,48 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+        assertEquals(-1, state.lastFetchTimestamp())
+        assertEquals(-1, state.lastCaughtUpTimestamp())
+      }
+      observerData.foreach { state =>
+        assertTrue(0 < state.replicaId)
+        assertTrue(0 < state.logEndOffset())
+        assertEquals(-1, state.lastFetchTimestamp())
+        assertEquals(-1, state.lastCaughtUpTimestamp())
+      }
```
nit: Should we just remove this block if `assertEquals(0, observerData.size)`?

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/QuorumInfo.java:
```diff
@@ -0,0 +1,153 @@
+    public String toString() {
+        return "QuorumInfo(" +
+            "leaderId=" + leaderId +
+            ", voters=" + voters.toString() +
+            ", observers=" + observers.toString() +
```
nit: I suppose that we could remove the `toString` as they as implicit.

dajac commented on core/src/test/scala/unit/kafka/server/DescribeQuorumRequestTest.scala:
```diff
@@ -54,30 +54,48 @@ class DescribeQuorumRequestTest(cluster: ClusterInstance) {
+      assertEquals(0, observerData.size)
+      voterData.foreach { state =>
+        assertTrue(0 < state.replicaId)
+        assertTrue(0 < state.logEndOffset())
+        assertEquals(-1, state.lastFetchTimestamp())
+        assertEquals(-1, state.lastCaughtUpTimestamp())
```
nit: We can also omit parenthesis for those..

dajac reviewed (commented):
@niket-goel Thanks for the update. Overall, LGTM. I left a few nits and one question.

niket-goel committed:
minor: some code styling fixes

niket-goel commented:
Thanks for the in-depth review @dajac @hachikuji and @dengziming. I have updated the PR with the final NITs from @dajac fixed. Are we good to merge with 1 LGTM or do we need 1 more?

> Also in response to _"Is it the expected behavior to have zero observers?"_

@dengziming also raised the same question in a previous comment, and we decided to tackle that in a separate PR. I will cut a JIRA for this.

dajac reviewed (approved):
LGTM. I will merge it when the build completes.

@niket-goel Could you update the KIP and notify the vote thread with the changes? Thanks.

hachikuji commented:
@dajac I think this is the reason there are no observers: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/raft/RaftManager.scala#L188. We should just use `config.nodeId` consistently regardless of `process.roles`. @niket-goel @dengziming Do either of you have time to fix this?

hachikuji reviewed (approved):
Thanks for all the iterations. LGTM.

niket-goel commented:
Thanks @hachikuji . I have created https://issues.apache.org/jira/browse/KAFKA-13986 for the observer count issue and will raise a PR once this one is merged

hachikuji commented:
@niket-goel It looks like we are running into https://issues.apache.org/jira/browse/KAFKA-13940, which is causing the new integration test to be flaky. I think our options are either to handle this error code in the client or to change the raft implementation to throw a retriable error (probably either `NotController` or `NotLeader`).

dajac commented:
> @niket-goel It looks like we are running into https://issues.apache.org/jira/browse/KAFKA-13940, which is causing the new integration test to be flaky. I think our options are either to handle this error code in the client or to change the raft implementation to throw a retriable error (probably either `NotController` or `NotLeader`).

I lean towards the server side change to be consistent with other APIs.

niket-goel committed:
Removing KRaftClusterTest Integration Test causing flakiness due to KAFKA-13940
